src/tools/crossgen2/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs

```diff
@@ -250,42 +250,44 @@ private void ProcessMarkStack()
-            PerfEventSource.Log.GraphProcessingStart();
-            if (_markingCompleted)
-                return;
-
-            do
+            using (StartStopEvents graphProcessingEvents = StartStopEvents.PerfEventSource.Log.GraphProcessingEvents())
             {
-                // Run mark stack algorithm as much as possible
-                PerfEventSource.Log.DependencyAnalysisStart();
-                ProcessMarkStack();
-                PerfEventSource.Log.DependencyAnalysisStop();
-
-                // Compute all dependencies which were not ready during the ProcessMarkStack step
-                ComputeDependencies(_deferredStaticDependencies);
-                foreach (DependencyNodeCore<DependencyContextType> node in _deferredStaticDependencies)
+                if (_markingCompleted)
+                    return;
+
+                do
                 {
-                    Debug.Assert(node.StaticDependenciesAreComputed);
-                    GetStaticDependenciesImpl(node);
-                }
+                    // Run mark stack algorithm as much as possible
+                    using (StartStopEvents dependencyAnalysisEvents = StartStopEvents.PerfEventSource.Log.DependencyAnalysisEvents())
+                    {
+                        ProcessMarkStack();
+                    }
 
-                _deferredStaticDependencies.Clear();
-            } while (_markStack.Count != 0);
+                    // Compute all dependencies which were not ready during the ProcessMarkStack step
+                    ComputeDependencies(_deferredStaticDependencies);
+                    foreach (DependencyNodeCore<DependencyContextType> node in _deferredStaticDependencies)
+                    {
+                        Debug.Assert(node.StaticDependenciesAreComputed);
+                        GetStaticDependenciesImpl(node);
+                    }
 
-            if (_resultSorter != null)
-                _markedNodes.Sort(_resultSorter);
+                    _deferredStaticDependencies.Clear();
+                } while (_markStack.Count != 0);
 
-            _markedNodesFinal = _markedNodes.ToImmutableArray();
-            _markedNodes = null;
-            _markingCompleted = true;
-            PerfEventSource.Log.GraphProcessingStop();
+                if (_resultSorter != null)
+                    _markedNodes.Sort(_resultSorter);
+
+                _markedNodesFinal = _markedNodes.ToImmutableArray();
+                _markedNodes = null;
+                _markingCompleted = true;
+            }
         }
 
         private bool AddToMarkStack(DependencyNodeCore<DependencyContextType> node, string reason, DependencyNodeCore<DependencyContextType> reason1, DependencyNodeCore<DependencyContextType> reason2)
         {
             if (_marker.MarkNode(node, reason1, reason2, reason))
             {
-                PerfEventSource.Log.AddedNodeToMarkStack();
+                StartStopEvents.PerfEventSource.Log.AddedNodeToMarkStack();
```

Comment:
I'm wondering: if we inverted the nesting (`StartStopEvents` nested under `PerfEventSource`) these lines would get shorter - would there be any drawbacks?

Reply:
Nope, that's definitely better (semantically and structurally); changed. I remember trying to do this yesterday and struggling a lot but it was actually really simple. I guess I just needed sleep :)